test/framework/src/main/java/org/elasticsearch/test/XContentTestUtils.java

```diff
@@ -123,4 +136,163 @@ private static String differenceBetweenObjectsIgnoringArrayOrder(String path, Ob
+     * <pre>
+     * {@code
+     *      (path) -> path.endsWith("foo1")
+     * }
+     * </pre>
+     *
+     * or if we don't want any random insertions in the "foo1" tree we could use
+     * <pre>
+     * {@code
+     *      (path) -> path.contains("foo1")
+     * }
+     * </pre>
+     */
+    public static BytesReference insertRandomFields(XContentType contentType, BytesReference xContent, Predicate<String> excludeFilter,
+            Random random, NamedXContentRegistry registry) throws IOException {
+        List<String> insertPaths;
+
+        try (XContentParser parser = createParser(registry, xContent, contentType)) {
+            List<String> possiblePaths = XContentTestUtils.getInsertPaths(parser);
+            if (excludeFilter == null) {
+                insertPaths = possiblePaths;
+            } else {
+                insertPaths = new ArrayList<>();
+                possiblePaths.stream().filter(excludeFilter.negate()).forEach(insertPaths::add);
+            }
+        }
+        try (XContentParser parser = createParser(registry, xContent, contentType)) {
+            Supplier<Object> value = () -> {
+                if (random.nextBoolean()) {
+                    return RandomObjects.randomStoredFieldValues(random, contentType);
+                } else {
+                    if (random.nextBoolean()) {
+                        return Collections.singletonMap(randomAsciiOfLength(random, 10), randomAsciiOfLength(random, 10));
+                    } else {
+                        return Collections.singletonList(randomAsciiOfLength(random, 10));
+                    }
+                }
+            };
+            return XContentTestUtils
+                    .insertIntoXContent(contentType.xContent(), xContent, insertPaths, () -> randomAsciiOfLength(random, 10), value)
+                    .bytes();
+        }
+    }
+
+    /**
+     * This utility method takes an XContentParser and walks the xContent structure to find all
+     * possible paths to where a new object or array starts. This can be used in tests that add random
+     * xContent values to test parsing code for errors or to check their robustness against new fields.
+     *
+     * The path uses dot separated fieldnames and numbers for array indices, similar to what we do in
+     * {@link ObjectPath}.
+     *
+     * As an example, the following json xContent:
+     * <pre>
+     *     {
+     *         "foo" : "bar",
+     *         "foo1" : [ 1, { "foo2" : "baz" }, 3, 4]
+     *         "foo3" : {
+     *             "foo4" : {
+     *                  "foo5": "buzz"
+     *             }
+     *         }
+     *     }
+     * </pre>
+     *
+     * Would return the following list:
+     *
+     * <ul>
+     *  <li>"" (the empty string is the path to the root object)</li>
+     *  <li>"foo1.1"</li>
+     *  <li>"foo3</li>
+     *  <li>"foo3.foo4</li>
+     * </ul>
+     */
+    static List<String> getInsertPaths(XContentParser parser) throws IOException{
```

Comment:
we could potentially remove this method and create the stack where we currently call this method? I think then we don't even need the assertions?